File: python/test/partition/partition_test.py
#  Copyright 2017-2025 John Snow Labs
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
import os
import unittest
import pytest
from sparknlp.partition.partition import Partition


@pytest.mark.fast
class PartitionTextTesSpec(unittest.TestCase):

    def setUp(self):
        self.txt_directory = f"file:///{os.getcwd()}/../src/test/resources/reader/txt"

    def runTest(self):
        text_df = Partition(content_type = "text/plain").partition(self.txt_directory)
        text_file_df = Partition().partition(f"{self.txt_directory}/simple-text.txt")

        self.assertTrue(text_df.select("txt").count() > 0)
        self.assertTrue(text_file_df.select("txt").count() > 0)


@pytest.mark.fast
class PartitionWordTesSpec(unittest.TestCase):

    def setUp(self):
        self.word_directory = f"file:///{os.getcwd()}/../src/test/resources/reader/doc"

    def runTest(self):
        doc_df = Partition(content_type = "application/msword").partition(self.word_directory)
        doc_file_df = Partition().partition(f"{self.word_directory}/fake_table.docx")

        self.assertTrue(doc_df.select("doc").count() > 0)
        self.assertTrue(doc_file_df.select("doc").count() > 0)


@pytest.mark.fast
class PartitionExcelTesSpec(unittest.TestCase):

    def setUp(self):
        self.excel_directory = f"file:///{os.getcwd()}/../src/test/resources/reader/xls"

    def runTest(self):
        xls_df = Partition(content_type = "application/vnd.ms-excel").partition(self.excel_directory)
        xls_file_df = Partition().partition(f"{self.excel_directory}/vodafone.xlsx")

        self.assertTrue(xls_df.select("xls").count() > 0)
        self.assertTrue(xls_file_df.select("xls").count() > 0)


@pytest.mark.fast
class PartitionPowerPointTesSpec(unittest.TestCase):

    def setUp(self):
        self.ppt_directory = f"file:///{os.getcwd()}/../src/test/resources/reader/ppt"

    def runTest(self):
        ppt_df = Partition(content_type = "application/vnd.ms-powerpoint").partition(self.ppt_directory)
        ppt_file_df = Partition().partition(f"{self.ppt_directory}/fake-power-point.pptx")

        self.assertTrue(ppt_df.select("ppt").count() > 0)
        self.assertTrue(ppt_file_df.select("ppt").count() > 0)


@pytest.mark.fast
class PartitionEmailTesSpec(unittest.TestCase):

    def setUp(self):
        self.eml_directory = f"file:///{os.getcwd()}/../src/test/resources/reader/email"

    def runTest(self):
        eml_df = Partition(content_type = "message/rfc822").partition(self.eml_directory)
        eml_file_df = Partition().partition(f"{self.eml_directory}/test-several-attachments.eml")

        self.assertTrue(eml_df.select("email").count() > 0)
        self.assertTrue(eml_file_df.select("email").count() > 0)


@pytest.mark.fast
class PartitionHtmlTesSpec(unittest.TestCase):

    def setUp(self):
        self.html_directory = f"file:///{os.getcwd()}/../src/test/resources/reader/html"

    def runTest(self):
        html_df = Partition(content_type = "text/html").partition(self.html_directory)
        html_file_df = Partition().partition(f"{self.html_directory}/fake-html.html")

        self.assertTrue(html_df.select("html").count() > 0)
        self.assertTrue(html_file_df.select("html").count() > 0)


@pytest.mark.fast
class PartitionUrlTesSpec(unittest.TestCase):

    def runTest(self):
        url_df = Partition().partition("https://www.wikipedia.org")
        urls_df = Partition().partition(["https://www.wikipedia.org", "https://example.com/"])

        self.assertTrue(url_df.select("html").count() > 0)
        self.assertTrue(urls_df.select("html").count() > 0)


@pytest.mark.fast
class PartitionPdfTesSpec(unittest.TestCase):

    def setUp(self):
        self.html_directory = f"file:///{os.getcwd()}/../src/test/resources/reader/pdf"

    def runTest(self):
        pdf_df = Partition(content_type = "application/pdf").partition(self.html_directory)
        pdf_file_df = Partition().partition(f"{self.html_directory}/text_3_pages.pdf")

        self.assertTrue(pdf_df.select("text").count() > 0)
        self.assertTrue(pdf_file_df.select("text").count() > 0)

File: src/test/scala/com/johnsnowlabs/partition/PartitionTest.scala
/*
 * Copyright 2017-2025 John Snow Labs
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.johnsnowlabs.partition

import com.johnsnowlabs.nlp.util.io.ResourceHelper
import com.johnsnowlabs.reader.{ElementType, HTMLElement}
import com.johnsnowlabs.tags.{FastTest, SlowTest}
import org.apache.spark.sql.functions.col
import org.scalatest.flatspec.AnyFlatSpec

import scala.collection.mutable

class PartitionTest extends AnyFlatSpec {

  val txtDirectory = "src/test/resources/reader/txt"
  val wordDirectory = "src/test/resources/reader/doc"
  val excelDirectory = "src/test/resources/reader/xls"
  val powerPointDirectory = "src/test/resources/reader/ppt"
  val emailDirectory = "src/test/resources/reader/email"
  val htmlDirectory = "src/test/resources/reader/html"
  val pdfDirectory = "src/test/resources/reader/pdf"
  val xmlDirectory = "src/test/resources/reader/xml"

  "Partition" should "work with text content_type" taggedAs FastTest in {
    val textDf = Partition(Map("content_type" -> "text/plain")).partition(txtDirectory)

    assert(!textDf.select(col("txt").getItem(0)).isEmpty)
  }

  it should "identify text file" taggedAs FastTest in {
    val textDf = Partition().partition(s"$txtDirectory/simple-text.txt")

    assert(!textDf.select(col("txt").getItem(0)).isEmpty)
  }

  it should "work with word content_type" taggedAs FastTest in {
    val wordDf = Partition(Map("content_type" -> "application/msword")).partition(wordDirectory)

    assert(!wordDf.select(col("doc").getItem(0)).isEmpty)
  }

  it should "identify word file" taggedAs FastTest in {
    val wordDf = Partition().partition(s"$wordDirectory/fake_table.docx")

    assert(!wordDf.select(col("doc").getItem(0)).isEmpty)
  }

  it should "work with excel content_type" taggedAs FastTest in {
    val excelDf =
      Partition(Map("content_type" -> "application/vnd.ms-excel")).partition(excelDirectory)

    assert(!excelDf.select(col("xls").getItem(0)).isEmpty)
  }

  it should "identify excel file" taggedAs FastTest in {
    val excelDf = Partition().partition(s"$excelDirectory/vodafone.xlsx")

    assert(!excelDf.select(col("xls").getItem(0)).isEmpty)
  }

  it should "work with email content_type" taggedAs FastTest in {
    val emailDf = Partition(Map("content_type" -> "message/rfc822")).partition(emailDirectory)

    assert(!emailDf.select(col("email").getItem(0)).isEmpty)
  }

  it should "wok with email file" taggedAs FastTest in {
    val emailDf = Partition().partition(s"$emailDirectory/test-several-attachments.eml")

    assert(!emailDf.select(col("email").getItem(0)).isEmpty)
  }

  it should "work with powerpoint content_type" taggedAs FastTest in {
    val pptDf = Partition(Map("content_type" -> "application/vnd.ms-powerpoint"))
      .partition(powerPointDirectory)

    assert(!pptDf.select(col("ppt").getItem(0)).isEmpty)
  }

  it should "identify powerpoint file" taggedAs FastTest in {
    val pptDf = Partition().partition(s"$powerPointDirectory/fake-power-point.pptx")

    assert(!pptDf.select(col("ppt").getItem(0)).isEmpty)
  }

  it should "work with html content_type" taggedAs FastTest in {
    val htmlDf = Partition(Map("content_type" -> "text/html")).partition(htmlDirectory)

    assert(!htmlDf.select(col("html").getItem(0)).isEmpty)
  }

  it should "identify html file" taggedAs FastTest in {
    val htmlDf = Partition().partition(s"$htmlDirectory/fake-html.html")

    assert(!htmlDf.select(col("html").getItem(0)).isEmpty)
  }

  it should "work with an URL" taggedAs SlowTest in {
    val htmlDf = Partition().partition("https://www.wikipedia.org")

    assert(!htmlDf.select(col("html").getItem(0)).isEmpty)
  }

  it should "work with a set of URLS" taggedAs SlowTest in {
    val htmlDf =
      Partition().partitionUrls(Array("https://www.wikipedia.org", "https://example.com/"))

    assert(!htmlDf.select(col("html").getItem(0)).isEmpty)
  }

  it should "identify a PDF file" taggedAs FastTest in {
    val pdfDf = Partition().partition(s"$pdfDirectory/text_3_pages.pdf")

    assert(!pdfDf.select(col("pdf")).isEmpty)
  }

  it should "work with PDF content_type" taggedAs FastTest in {
    val pdfDf = Partition(Map("content_type" -> "application/pdf")).partition(pdfDirectory)

    assert(!pdfDf.select(col("pdf")).isEmpty)
  }

  it should "work with text in memory" taggedAs FastTest in {
    import ResourceHelper.spark.implicits._
    val content =
      """
        |The big brown fox
        |was walking down the lane.
        |
        |At the end of the lane,
        |the fox met a bear.
        |""".stripMargin

    val textDf = Partition(Map("groupBrokenParagraphs" -> "true")).partitionText(content)

    val elements: Seq[HTMLElement] = textDf
      .select("txt")
      .as[Seq[HTMLElement]]
      .collect()
      .head

    val expectedElements = Seq(
      HTMLElement(
        ElementType.NARRATIVE_TEXT,
        "The big brown fox was walking down the lane.",
        mutable.Map("paragraph" -> "0")),
      HTMLElement(
        ElementType.NARRATIVE_TEXT,
        "At the end of the lane, the fox met a bear.",
        mutable.Map("paragraph" -> "0")))

    assert(elements.head.elementType == expectedElements.head.elementType)
    assert(elements.head.content == expectedElements.head.content)
  }

  it should "work with XML content_type" taggedAs FastTest in {
    val pdfDf = Partition(Map("content_type" -> "application/xml")).partition(xmlDirectory)

    assert(!pdfDf.select(col("xml")).isEmpty)
  }

}
